src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -336,7 +335,13 @@ private void SendReset()
                 // Don't send a RST_STREAM if we've already received one from the server.
                 if (_resetException == null)
                 {
-                    _connection.LogExceptions(_connection.SendRstStreamAsync(StreamId, Http2ProtocolErrorCode.Cancel));
+                    // Use dedicated lock to prevent sending a frame with EndStream flag after an RST_STREAM frame.
+                    // If execution reached this line, it's guaranteed that
+                    // _requestCompletionState == StreamCompletionState.Failed or _responseCompletionState == StreamCompletionState.Failed
+                    lock (SyncObject)
```

Comment:
We definitely don't want to do this as it will result in lock re-entrancy, which is bad.

That's why we generally do `Debug.Assert(!Monitor.IsEntered(SyncObject));` before taking the lock.

Instead we should ensure callers hold the lock around calling this routine. We may need to do this anyway in order to prevent races fully.

Reply:
Ok, makes sense. I'm not happy with a lock re-entrancy here, either, but was not sure if getting rid of it justifies adding `lock` blocks to all callers. Will do it now.

Reply:
Locking is moved to SendReset callers.